### PR TITLE
test: add unit tests to OIDC middleware

### PR DESCRIPTION
**Pull request by edwin-Marrima**
## Description

closes: #1395 

## Review Checklist
- [x ] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x ] The correct base branch is being used, if not `main`
- [x ] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: internal/graph/check.go [hunk 1]
```diff
@@ -57,7 +57,6 @@ func (r *ResolveCheckResponse) GetResolutionMetadata() *ResolutionMetadata {
 	if r != nil {
 		return r.ResolutionMetadata
 	}
-
```
Comment (jon-whit) [inline internal/graph/check.go:60]: Unrelated code change. To keep this PR history clean, would you mind reverting this change please?

File: internal/authn/oidc/oidc.go [hunk 2]
```diff
@@ -118,14 +120,13 @@ func (oidc *RemoteOidcAuthenticator) Authenticate(requestContext context.Context
 	return principal, nil
 }
 
-func (oidc *RemoteOidcAuthenticator) fetchKeys() error {
+func fetchKeys(oidc *RemoteOidcAuthenticator) error {
```
Comment (jon-whit) [inline internal/authn/oidc/oidc.go:123]: Let's just replace/rename this to `fetchJWKs` if we're going to alias it above.

File: internal/authn/oidc/oidc_test.go [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+	"github.com/openfga/openfga/internal/authn"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc/metadata"
+	"log"
+	"strings"
+	"testing"
+)
+
+func TestRemoteOidcAuthenticator_Authenticate(t *testing.T) {
+	t.Run("When the authorization header is missing from the gRPC metadata of the request, returns 'missing bearer token' error.", func(t *testing.T) {
```
Comment (jon-whit) [inline internal/authn/oidc/oidc_test.go:18]: Use `_` instead of whitespace please. It makes it easier to search/find since Go replaces whitespace with underscores in `go test` output.

File: internal/authn/oidc/oidc_test.go [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+		//given
```
Comment (jon-whit) [inline internal/authn/oidc/oidc_test.go:19]: Can we drop these noisy comments. They are intuitive given the test layout anyways. You did a good job making the tests easy to distinguish to begin with.

File: internal/authn/oidc/oidc_test.go [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+					jwt.MapClaims{
+						"iss": "wrong_issuer",
+					},
+					nil,
+				)
+			},
+			expectedError: "invalid issuer",
+		},
+		{
+			testDescription: "when_token's_audience_does_not_match_the_one_provided_in_the_server_configuration,_MUST_return_'invalid_audience'_error",
```
Comment (miparnisari) [inline internal/authn/oidc/oidc_test.go:60]: why does this test not have `expectedError: "invalid audience"`?

File: internal/authn/oidc/oidc_test.go [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+	"testing"
+)
+
+func TestRemoteOidcAuthenticator_Authenticate(t *testing.T) {
+	t.Run("when_the_authorization_header_is_missing_from_the_gRPC_metadata_of_the_request,_returns_'missing_bearer_token'_error", func(t *testing.T) {
+		authenticator := &RemoteOidcAuthenticator{}
+		_, err := authenticator.Authenticate(context.Background())
+		require.Equal(t, authn.ErrMissingBearerToken, err)
+	})
+	testCases := []struct {
```
Comment (miparnisari) [inline internal/authn/oidc/oidc_test.go:23]: please add two more test cases
- a token with `exp` (expiration date) in the past should be rejected
- a token with `iat` (issued at) in the future should be rejected

**Review by miparnisari — CHANGES_REQUESTED** [review]

File: internal/authn/oidc/oidc_test.go [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+}
+
+// quickConfigSetup sets up a basic configuration for testing purposes.
+func quickConfigSetup(jwkKid, jwtKid, issuerURL, audience string, jwtClaims jwt.MapClaims, privateKeyOverride *rsa.PrivateKey) (*RemoteOidcAuthenticator, context.Context, error) {
+	// Generate JWT signature keys
+	privateKey, publicKey := generateJWTSignatureKeys()
+	if privateKeyOverride != nil {
+		privateKey = privateKeyOverride
+	}
+	// assign mocked JWKS fetching function to global function
```
Comment (miparnisari) [inline internal/authn/oidc/oidc_test.go:142]: yikes we are using a global function 😱 

are sure that by virtue of having this test we are not overriding the production code? can you do a sanity check and post the result in the PR description?

1. run a local JWK server (a basic server that responds to `.well-known/openid-configuration` with a static file)
2. run `./openfga run --playground-enabled=false --authn-method=oidc --authn-oidc-audience=local --authn-oidc-issuer="http://LOCAL-ADDRESS:LOCAL-PORT"`
4. make a call to OpenFGA's Check API with a token that has a token signed with a different key, you should get "invalid bearer token"

File: internal/authn/oidc/oidc_test.go [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+					nil,
+				)
+			},
+			expectedError: "invalid subject",
+		},
+	}
+
+	for _, testC := range testCases {
+		t.Run(testC.testDescription, func(t *testing.T) {
+			oidc, requestContext, err := testC.testSetup()
```
Comment (miparnisari) [inline internal/authn/oidc/oidc_test.go:97]:
```suggestion
		t.Run(testC.testDescription, func(t *testing.T) {
		        if testC.expectedError == "" {
		              t.Fail("this suite is to test error cases and this test didn't have an error")
		        } 
			oidc, requestContext, err := testC.testSetup()
```

otherwise line 99 will silently pass

File: internal/authn/oidc/oidc_test.go (hunk 7 above)
Comment (edwin-Marrima) [reply internal/authn/oidc/oidc_test.go:185]: @miparnisari could you please clarify what exactly you want to validate with sanity check?

Comment (miparnisari) [reply internal/authn/oidc/oidc_test.go:185]: i want to be 100% sure that by doing `fetchJWKs = fetchJWK` we are not overriding the production code with test code

Comment (edwin-Marrima) [reply internal/authn/oidc/oidc_test.go:185]: If this is the purpose, I believe that a sanity test that only checks whether a token is valid is more than enough. 

Given that the tests solely simulate the `fetchJWK` behavior, and as per its design, this function exclusively retrieves JWKs without conducting any validation, I perceive no potential for altering the validation behavior unless there's an oversight on my part.           

Here is the sanity test that performs token validation: 

`make build && ./openfga run --playground-enabled=false  --authn-method=oidc --authn-oidc-issuer="http://localhost:8185/realms/internal"  --authn-oidc-audience="account"`
image

Request with valid token:
image

Request with invalid token: 
image

**Review by miparnisari — DISMISSED** [review]

**Comment by miparnisari** [comment]
@edwin-Marrima please run `make lint` and fix any outstanding issues 🙏

**Review by miparnisari — APPROVED** [review]
